resolwe_runtime_utils.py

```diff
@@ -48,51 +137,168 @@ def _get_json(value):
+            )
 
-    JSON object is of the form:
-    { key: [values[0], values[1], ... ] },
-    where values represent the given list of parameters.
+        if response["type_data"] != "OK":
+            return False
+        return True
+
+    message = json.dumps(data).encode("utf-8")
+    message_length = "{length:0{size}d}".format(
+        length=len(message), size=size_bytes
+    ).encode("utf-8")
+    bytes_to_send = message_length + message
+    s.sendall(bytes_to_send)
+    response = _receive_data(s)
+    if not _check_response(response):
+        logger.error("Error in respone to %s: %s.", data, response)
+        raise RuntimeError("Wrong response received, terminating processing.")
+
+
+def send_json_message(data):
+    """Send data over socket."""
+
+    @_retry(max_retries=5)
+    def _connect(socket_path: str) -> socket.SocketType:
+        """Connect socket."""
+        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        sock.connect(socket_path)
+        sock.settimeout(SOCKET_TIMEOUT)
+        return sock
 
-    """
-    return json.dumps({key: [_get_json(value) for value in values]})
+    try:
+        sock = _connect(str(COMMUNICATOR_SOCKET))
+        send_data(sock, data)
+    finally:
+        with suppress(Exception):
+            sock.close()
 
 
-def annotate_entity(key, value):
+def _copy_file_or_dir(entries):
+    """Copy file and all its references to data_volume.
+
+    The entry is a path relative to the DATA_LOCAL_VOLUME (our working
+    directory). It must be copied to the DATA_VOLUME on the shared
+    filesystem.
+    """
+    for entry in entries:
+        source = Path(entry)
+        destination = DATA_VOLUME / source
+        if not destination.parent.is_dir():
+            destination.parent.mkdir(parents=True)
+
+        if source.is_dir():
+            if not destination.exists():
+                shutil.copytree(str(source), str(destination))
+            else:
+                # If destination directory exists the copytree will fail.
+                # In such case perform a recursive call with entries in
+                # the source directory as arguments.
+                #
+                # TODO: fix when we support Python 3.8 and later. See
+                # dirs_exist_ok argument to copytree method.
+                _copy_file_or_dir(source.glob("*"))
+        elif source.is_file():
+            if not destination.parent.is_dir():
+                destination.parent.mkdir(parents=True)
+            # Use copy2 to preserve file metadata, such as file creation
+            # and modification times.
+            shutil.copy2(str(source), str(destination))
+
+
+def _copy_if_needed(data):
+    """Check if data output represents file.
+
+    And copy the file to the shared filesystem if this is the case.
     """
-    Convert the given annotation to a JSON object.
+    if isinstance(data, dict):
+        if "file" in data and not "file_temp" in data:
```

Comment:
Is there a case when the output contains `file_temp`?

Reply:
I was not sure. Most probably not in this case so I will remove the condition.